### PR TITLE
Prints benchmark results in a neat table and attempts to run every benchmark 

### LINKED ISSUE #1440: Visualize benchmarks in a single table view
### Feature description
Display benchmark results as a single table in the console

### Feature motivation
It's pain to see and scroll through results across combinations of benchmark types and backends.

### PR DESCRIPTION
**Pull request by akhildevelops**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

Fixes #1440

### Changes

**Benchmark results will be displayed as below based on benchmarks X backends**
image

**Benchmark process will continue to run even if there's a failure in any one of the combination of benchmarks and backends**
image

### Testing

There's a test case included and ran this feature across all the available benchmarks and backends.

### CODE REVIEW

**Comment by akhildevelops** [comment]
Is the code ok to get merged ?

**Review by antimora — COMMENTED** [review]
I haven't reviewed closely - leaving it up to @syl20bnr but I had small suggestion. It'd be convenient if we make the [table Markdown compatible](https://www.codecademy.com/resources/docs/markdown/tables) so we can copy the table output into any MD renderer (Discord, Github, other docs).

**Comment by akhildevelops** [comment]
Intersting idea, that can be done.  :handshake:

**Comment by akhildevelops** [comment]
```
❯ cargo run --release --bin burnbench -- run --benches unary --backends wgpu candle-cuda tch-cpu        
    Finished release [optimized] target(s) in 0.22s
     Running `target/release/burnbench run --benches unary --backends wgpu candle-cuda tch-cpu`
Executing the following benchmark and backend combinations (Total: 3):
Running benchmarks...
==========Benchmarking unary on wgpu==========
    Finished bench [optimized] target(s) in 0.14s
     Running benches/unary.rs (target/benchmarks/release/deps/unary-14750bb533c606ad)

        Timestamp: 1710384072278
        Git Hash: 24db2c25023771b5d4a856734457f232dcafcc91
        Benchmarking - unary
―――――――― Result ―――――――――
  Samples     10
  Mean        1.660ms
  Variance    81.000ns
  Median      1.545ms
  Min         1.380ms
  Max         2.271ms
―――――――――――――――――――――――――
        
==========Benchmarking unary on candle-cuda==========
    Finished bench [optimized] target(s) in 0.15s
     Running benches/unary.rs (target/benchmarks/release/deps/unary-216654a32a33583a)

        Timestamp: 1710384072664
        Git Hash: 24db2c25023771b5d4a856734457f232dcafcc91
        Benchmarking - unary
―――――――― Result ―――――――――
  Samples     10
  Mean        8.667ms
  Variance    23.000ns
  Median      8.611ms
  Min         8.488ms
  Max         8.981ms
―――――――――――――――――――――――――
        
==========Benchmarking unary on tch-cpu==========
    Finished bench [optimized] target(s) in 0.15s
     Running benches/unary.rs (target/benchmarks/release/deps/unary-26a7ed056fa878c2)

        Timestamp: 1710384073296
        Git Hash: 24db2c25023771b5d4a856734457f232dcafcc91
        Benchmarking - unary
―――――――― Result ―――――――――
  Samples     10
  Mean        15.416ms
  Variance    34.880µs
  Median      13.681ms
  Min         8.867ms
  Max         24.696ms
―――――――――――――――――――――――――
        
==========Benchmark Results==========

| Benchmark      | Backend                            | Runtime        |
|----------------|------------------------------------|----------------|
| unary          | jit<wgpu>-BestAvailable            | 1.660ms        |
| unary          | candle-Cuda(0)                     | 8.666ms        |
| unary          | tch-Cpu                            | 15.415ms       |
```

**Comment by akhildevelops** [comment]
Below is the rendering in this github's comment box
        
==========Benchmark Results==========

| Benchmark      | Backend                            | Runtime        |
|----------------|------------------------------------|----------------|
| unary          | jit<wgpu>-BestAvailable            | 1.660ms        |
| unary          | candle-Cuda(0)                     | 8.666ms        |
| unary          | tch-Cpu                            | 15.415ms       |

File: backend-comparison/src/burnbenchapp/base.rs [hunk 1]
```diff
@@ -246,7 +253,36 @@ pub(crate) fn run_backend_comparison_benchmarks(
...
+            } else {
+                println!("Cannot find the benchmark-record file: {}", file_path);
+            };
+        }
+        fs::remove_file(result_path).ok();
+    } else {
+        println!("Cannot find the result file: backend-comparison/tmp_bench.json");
+    };
+    println!("{}Benchmark Results{}\n", filler, filler);
+    println!("{}", benchmark_results);
```
Comment (syl20bnr) [inline backend-comparison/src/burnbenchapp/base.rs:287]: Create a function for printing the results.

File: backend-comparison/src/persistence/base.rs [hunk 2]
```diff
@@ -77,10 +77,23 @@ pub fn save<B: Backend>(
...
         );
         let file_path = cache_dir.join(file_name);
-        let file = fs::File::create(file_path).expect("Benchmark file should exist or be created");
+        let file =
+            fs::File::create(file_path.clone()).expect("Benchmark file should exist or be created");
         serde_json::to_writer_pretty(file, &record)
             .expect("Benchmark file should be updated with benchmark results");
 
+        // Append the benchmark result filepath in a temp file to be later picked by benchrun
+        let curdir_filepath = env::current_dir()
```
Comment (syl20bnr) [inline backend-comparison/src/persistence/base.rs:86]: Not sure what the current directory is. Let's use the cache directory for the temporary results file.

File: backend-comparison/src/persistence/base.rs [hunk 3]
```diff
@@ -154,3 +167,124 @@ impl Serialize for BenchmarkRecord {
...
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: burn::serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_map(BenchmarkRecordVisitor)
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct BenchMarkCollection {
```
Comment (syl20bnr) [inline backend-comparison/src/persistence/base.rs:233]: BenchMark -> Benchmark

File: backend-comparison/src/persistence/base.rs [hunk 4]
```diff
@@ -154,3 +167,124 @@ impl Serialize for BenchmarkRecord {
...
+            f,
+            "| {0:<15}| {1:<35}| {2:<15}|\n|{3:-<16}|{4:-<36}|{5:-<16}|",
+            "Benchmark", "Backend", "Runtime", "", "", ""
+        )?;
+        for record in self.records.iter() {
+            let backend = [record.backend.clone(), record.device.clone()].join("-");
+            writeln!(
+                f,
+                "| {0:<15}| {1:<35}| {2:<15.3?}|",
+                record.results.name, backend, record.results.computed.mean
```
Comment (syl20bnr) [inline backend-comparison/src/persistence/base.rs:249]: Use the median instead of mean.

File: backend-comparison/src/persistence/base.rs [hunk 5]
```diff
@@ -154,3 +167,124 @@ impl Serialize for BenchmarkRecord {
...
+pub(crate) struct BenchMarkCollection {
+    pub records: Vec<BenchmarkRecord>,
+}
+
+impl Display for BenchMarkCollection {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        writeln!(
+            f,
+            "| {0:<15}| {1:<35}| {2:<15}|\n|{3:-<16}|{4:-<36}|{5:-<16}|",
+            "Benchmark", "Backend", "Runtime", "", "", ""
```
Comment (syl20bnr) [inline backend-comparison/src/persistence/base.rs:242]: Runtime -> Median

File: backend-comparison/src/persistence/base.rs [hunk 6]
```diff
@@ -154,3 +167,124 @@ impl Serialize for BenchmarkRecord {
...
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn get_benchmark_result() {
+        let sample_result = r#"{"backend":"candle","device":"Cuda(0)","gitHash":"02d37011ab4dc773286e5983c09cde61f95ba4b5","name":"unary","max":8858,"mean":8629,"median":8592,"min":8506,"numSamples":10,"options":null,"rawDurations":[{"secs":0,"nanos":8858583},{"secs":0,"nanos":8719822},{"secs":0,"nanos":8705335},{"secs":0,"nanos":8835636},{"secs":0,"nanos":8592507},{"secs":0,"nanos":8506423},{"secs":0,"nanos":8534337},{"secs":0,"nanos":8506627},{"secs":0,"nanos":8521615},{"secs":0,"nanos":8511474}],"shapes":[[32,512,1024]],"timestamp":1710208069697,"variance":0}"#;
```
Comment (syl20bnr) [inline backend-comparison/src/persistence/base.rs:263]: Pretty print the sample so that it is more readable.

**Comment by akhildevelops** [comment]
Hi @syl20bnr, I've made changes according to the suggestions you've provided.

**Review by syl20bnr — APPROVED** [review]
Thank you !

**Comment by syl20bnr** [comment]
@akhildevelops I forget to tell you in case you did not know it that you should not use the main branch of your fork when you submit a pull request and instead you should create a dedicate branch for each PR. This way you can more easily work on multiple PRs at once and keep the main branch up to date with the upstream repository.

**Comment by akhildevelops** [comment]
Sure Thanks :handshake:

**Pull request merged**